src/debug/tester/tester.c

```diff
@@ -0,0 +1,255 @@
+//SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2025 Intel Corporation. All rights reserved.
+//
+// Author: Marcin Szkudlinski <marcin.szkudlinski@linux.intel.com>
+
+#include <stddef.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <rtos/init.h>
+#include <sof/lib/uuid.h>
+#include <sof/audio/component.h>
+#include <sof/audio/module_adapter/module/generic.h>
+#include <sof/audio/sink_api.h>
+#include <sof/audio/source_api.h>
+#include <sof/audio/sink_source_utils.h>
+#include "tester.h"
+#include "tester_dummy_test.h"
+
+/**
+ * Tester module is a framework for a runtime testing that need a special test code
+ * injected into the SOF system, i.e.
+ *  - provide an extra load to CPU by running additional thread
+ *  - execute some incorrect operations
+ *
+ * The idea is to allow using of special cases in testing like Continuous Integration or pre-release
+ * testing using a standard, production build.
+ *
+ * In order to have only one testing module (in meaning of a SOF module with separate UUID),
+ * a framework is introduced, where a test to be executed is selected by IPC parameter
+ *
+ * In production, module should be compiled as a loadable module, as it should not needed to be
+ * loaded and used on customers' boards.
+ * During developing, however, the module may be built in to keep debugging simpler
+ *
+ */
+
+LOG_MODULE_REGISTER(tester, CONFIG_SOF_LOG_LEVEL);
+
+SOF_DEFINE_REG_UUID(tester);
+
+DECLARE_TR_CTX(tester_tr, SOF_UUID(tester_uuid), LOG_LEVEL_INFO);
+
+struct tester_init_config {
+	struct ipc4_base_module_cfg ipc4_cfg;
+	int32_t test_type;
+} __attribute__((packed, aligned(4)));
+
+static int tester_init(struct processing_module *mod)
+{
+	struct module_data *md = &mod->priv;
+	struct comp_dev *dev = mod->dev;
+	struct module_config *cfg = &md->cfg;
+	size_t bs = cfg->size;
+	struct tester_module_data *cd = NULL;
+	int ret = 0;
+
+	if (bs != sizeof(struct tester_init_config)) {
+		comp_err(dev, "Invalid config");
+		return -EINVAL;
+	}
+
+	/* allocate ctx for test module in shared memory - to allow all non-standard operations
+	 * without problems with cache
+	 */
+	cd = rzalloc(SOF_MEM_ZONE_SYS_SHARED, 0, SOF_MEM_CAPS_RAM, sizeof(*cd));
+	if (!cd) {
+		comp_err(dev, "Out of memory");
+		return -ENOMEM;
+	}
+
+	struct tester_init_config *init_cfg = (struct tester_init_config *)cfg->init_data;
+
+	cd->test_case_type = init_cfg->test_type;
+	switch (cd->test_case_type) {
+	case TESTER_MODULE_CASE_DUMMY_TEST:
+		cd->tester_case_interface = &tester_interface_dummy_test;
+		break;
+
+	default:
+		comp_err(dev, "Invalid config, unknown test type");
+		rfree(cd);
+		return -EINVAL;
+	}
+
+	module_set_private_data(mod, cd);
+
+	if (cd->tester_case_interface->init)
+		ret = cd->tester_case_interface->init(mod, &cd->test_case_ctx);
+
+	if (ret) {
```

Comment:
nitpick: I'd move this under the `if` in line 88 above - `ret` can become non-zero only there.